### PR TITLE
Structure

### PR DESCRIPTION
**Pull request by lecorguille**
Here is a tool which belong to the population genetics thematic

@yvanlebras Finally, I started from scratch.

### CODE REVIEW

File: tools/structure/structure.xml [hunk 1]
```diff
@@ -0,0 +1,491 @@
...
+        cp '$extraparams' '$out_extraparams' &&
+
+        mkdir out log &&
+
+        #for $run in range(1, (int($nb_run) + 1)):
+            structure -i '$infile' -o outfile -m '$mainparams' -e '$extraparams' > 'log/run${run}_K_${main.MAXPOPS}.log' &&
+            mv 'outfile_f' 'out/run${run}_K_${main.MAXPOPS}.out' &&
+        #end for
+
+        echo ""
```
Comment (bgruening) [inline tools/structure/structure.xml]: ?

File: tools/structure/structure.xml [hunk 2]
```diff
@@ -0,0 +1,491 @@
...
+
+
+            <!-- MISCELLANEOUS -->
+
+            <param argument="COMPUTEPROB" checked="true" type="boolean" label="Estimate the probability of the Data under the model." help="This is used when choosing the best number of subpopulations." truevalue="1" falsevalue="0" />
+            <param argument="ADMBURNIN" value="500" type="integer" label="[only relevant for linkage model]: Initial period of burnin with admixture model (see Readme)" />
+            <param argument="ALPHAPROPSD" value="0.025" type="float" label="SD of proposal for updating alpha" />
+            <param argument="STARTATPOPINFO" checked="false" type="boolean" label="Use given populations as the initial condition for population origins." help="(Need POPDATA==1). It is assumed that the PopData in the input file are between 1 and k where k is less or equal MAXPOPS." truevalue="1" falsevalue="0" />
+            <conditional name="randomize_cond">
+                <param argument="RANDOMIZE" checked="true" type="boolean" label="use new random seed for each run" truevalue="1" falsevalue="0" />
```
Comment (bgruening) [inline tools/structure/structure.xml]: can we make this a select, following our guidelines?

File: tools/structure/structure.xml [hunk 3]
```diff
@@ -0,0 +1,491 @@
...
+                <when value="0">
+                    <param argument="SEED" value="2245" type="integer" label="seed value for random number generator" help="(must set RANDOMIZE=0)" />
+                </when>
+            </conditional>
+            <param argument="METROFREQ" value="10" type="integer" label="Frequency of using Metropolis step to update Q under admixture model" help="(ie use the metr. move every i steps).  If this is set to 0, it is never used. (Proposal for each q^(i) sampled from prior. The goal is to improve mixing for small alpha.)" />
+            <param argument="REPORTHITRATE" checked="false" type="boolean" label="report hit rate if using METROFREQ" truevalue="1" falsevalue="0" />
+        </section>
+    </inputs>
+    <outputs>
+        <data name="out_mainparams" format="txt" label="run_K_${main.MAXPOPS}.mainparams" />
```
Comment (bgruening) [inline tools/structure/structure.xml]: is this really needed as output? Or can we write this to stdout?
Maybe we can make it optional, as well as the logs?

File: tools/structure/structure.xml (hunk 1 above)
Comment (lecorguille) [reply tools/structure/structure.xml]: I didn’t want to finish with a &&

File: tools/structure/structure.xml (hunk 3 above)
Comment (lecorguille) [reply tools/structure/structure.xml]: All outputs are needed by my user. 
The param for archive and the log should be used to draw statitics ... one day

File: tools/structure/structure.xml (hunk 2 above)
Comment (lecorguille) [reply tools/structure/structure.xml]: No problemo

Comment (lecorguille) [reply tools/structure/structure.xml]: Done

File: tools/structure/structure.xml (hunk 1 above)
Comment (nsoranzo) [reply tools/structure/structure.xml]: You can do something like:
```
        mkdir out log

        #for $run in range(1, (int($nb_run) + 1)):
            && structure -i '$infile' -o outfile -m '$mainparams' -e '$extraparams' > 'log/run${run}_K_${main.MAXPOPS}.log'
            && mv 'outfile_f' 'out/run${run}_K_${main.MAXPOPS}.out'
        #end for
```

Comment (lecorguille) [reply tools/structure/structure.xml]: Indeed ^^'

Comment (lecorguille) [reply tools/structure/structure.xml]: Done

File: tools/structure/structure.xml [hunk 4]
```diff
@@ -0,0 +1,504 @@
+<tool id="structure" name="Structure" version="2.3.4.1">
```
Comment (bgruening) [inline tools/structure/structure.xml]: why is this version .1

File: tools/structure/structure.xml (hunk 3 above)
Comment (bgruening) [reply tools/structure/structure.xml]: Makes, sense but can we make then optional?

File: tools/structure/structure.xml (hunk 4 above)
Comment (lecorguille) [reply tools/structure/structure.xml]: `2.3.4` is the Structure version
`.1` is the wrapper version for this Structure version

File: tools/structure/structure.xml [hunk 5]
```diff
@@ -0,0 +1,504 @@
...
+    <description>using multi-locus genotype data to investigate population structure</description>
+    <requirements>
+        <requirement type="package" version="2.3.4">structure</requirement>
+    </requirements>
+    <command detect_errors="exit_code"><![CDATA[
+        cp '$mainparams' '$out_mainparams' &&
```
Comment (yhoogstrate) [inline tools/structure/structure.xml]: This feels a bit redundant @lecorguille . Would it be an idea to `mv '$mainparams' '$out_mainparams' &&` and use  `$out_mainparams` elsewhere? (same for extraparams)

File: tools/structure/structure.xml [hunk 6]
```diff
@@ -0,0 +1,504 @@
...
+        cp '$extraparams' '$out_extraparams' &&
+
+        mkdir out log
+
+        #for $run in range(1, (int($nb_run) + 1)):
```
Comment (yhoogstrate) [inline tools/structure/structure.xml]: `range(1, int($nb_run) + 1):`

File: tools/structure/structure.xml [hunk 7]
```diff
@@ -0,0 +1,504 @@
...
+            <param argument="RECESSIVEALLELES" checked="false" type="boolean" label="Data file contains dominant markers (eg AFLPs) and a row to indicate which alleles are recessive" truevalue="1" falsevalue="0" />
+            <param argument="MAPDISTANCES" checked="false" type="boolean" label="Data file contains row of map distances between loci" truevalue="1" falsevalue="0" />
+
+
+            <!--Advanced data file options-->
+
+            <param argument="PHASED" checked="false" type="boolean" label="Data are in correct phase (relevant for linkage model only)" truevalue="1" falsevalue="0" />
+            <param argument="PHASEINFO" checked="false" type="boolean" label="The data for each individual contains a line indicating phase (linkage model)" truevalue="1" falsevalue="0" />
+            <param argument="MARKOVPHASE" checked="false" type="boolean" label="The phase info follows a Markov model." truevalue="1" falsevalue="0" />
+            <param argument="NOTAMBIGUOUS" value="-999" type="integer" label="For use in some analyses of polyploid data" />
```
Comment (yhoogstrate) [inline tools/structure/structure.xml]: The label is a bit vague

File: tools/structure/structure.xml [hunk 8]
```diff
@@ -0,0 +1,504 @@
...
+        </section>
+        <section name="extra" title="extraparams" expanded="False">
+
+            <param argument="NOADMIX" checked="false" type="boolean" label="Use no admixture model" help="(0/True=admixture model, 1/False=no-admix)" truevalue="1" falsevalue="0" />
```
Comment (yhoogstrate) [inline tools/structure/structure.xml]: Would it make sense to switch this to "Use admixture model" and put it be default on true?

File: tools/structure/structure.xml [hunk 9]
```diff
@@ -0,0 +1,504 @@
...
+                    <param argument="POPSPECIFICLAMBDA" checked="false" type="boolean" label="Infer a separate lambda for each pop" help="(only if INFERLAMBDA=1/True)." truevalue="1" falsevalue="0" />
+                </when>
+            </conditional>
+            <param argument="LAMBDA" value="1.0" type="float" label="Dirichlet parameter for allele frequencies" />
+
+
+            <!-- PRIORS -->
+
+            <param argument="FPRIORMEAN" value="0.01" type="float" label="Prior mean and SD of Fst for pops." />
+            <param argument="FPRIORSD" value="0.05" type="float" label="The prior is a Gamma distribution with these parameters" />
```
Comment (yhoogstrate) [inline tools/structure/structure.xml]: I suppose this refers to a standard deviation (SD?). If so, please add this to the label

File: tools/structure/structure.xml [hunk 10]
```diff
@@ -0,0 +1,504 @@
+<tool id="structure" name="Structure" version="2.3.4.1">
+    <description>using multi-locus genotype data to investigate population structure</description>
+    <requirements>
+        <requirement type="package" version="2.3.4">structure</requirement>
+    </requirements>
+    <command detect_errors="exit_code"><![CDATA[
```
Comment (yhoogstrate) [inline tools/structure/structure.xml]:
```
<version_command><![CDATA[
structure | grep -E -o 'Version.+'
]]></version_command>
```

Comment (lecorguille) [reply tools/structure/structure.xml]: Thanks
Done

File: tools/structure/structure.xml (hunk 9 above)
Comment (lecorguille) [reply tools/structure/structure.xml]: Done

File: tools/structure/structure.xml (hunk 7 above)
Comment (lecorguille) [reply tools/structure/structure.xml]: Done

File: tools/structure/structure.xml (hunk 6 above)
Comment (lecorguille) [reply tools/structure/structure.xml]: Done

File: tools/structure/structure.xml (hunk 5 above)
Comment (lecorguille) [reply tools/structure/structure.xml]: Done

File: tools/structure/structure.xml (hunk 8 above)
Comment (lecorguille) [reply tools/structure/structure.xml]: Hum, I don't want to change this label even if indeed it's odd :/

**Comment by lecorguille** [comment]
@yhoogstrate @bgruening Thanks a lot for your review.
@yhoogstrate To be honest, I'm sure you spend more time than me to read and understand those different parameters. 🙇

**Comment by bgruening** [comment]
@lecorguille thanks for addressing our comments.
@yhoogstrate ok?

**Comment by yhoogstrate** [comment]
Yep, we can go. Thanks all 👍

**Comment by lecorguille** [comment]
Thanks to both of you for the review

**Pull request merged**